### PR TITLE
Simplified allocation of heap buffer during gateway fetch on win32.

### PR DESCRIPTION
**Pull request by Dimi1010**
- Replaced the `buffer` and `buffer2` raw memory allocations with a managed vector.
- Removed the need of `buffer2` via resizing the vector instead.
- `GetAdaptersInfo` will no longer be called twice if the first call does not experience buffer overflow.

### CODE REVIEW

File: Pcap++/src/PcapLiveDevice.cpp [hunk 1]
```diff
@@ -982,16 +983,18 @@ void PcapLiveDevice::setDefaultGateway()
...
 	retVal = GetAdaptersInfo(adapterInfo, &outBufLen);
-	uint8_t* buffer2 = new uint8_t[outBufLen];
 	if (retVal == ERROR_BUFFER_OVERFLOW)
-		adapterInfo = (IP_ADAPTER_INFO *)buffer2;
-
-	retVal = GetAdaptersInfo(adapterInfo, &outBufLen);
+	{
+		buffer.resize(outBufLen);
+		// Repins the adapter info pointer to the vector data pointer as the vector might be reallocated during the resize.
+		adapterInfo = reinterpret_cast<IP_ADAPTER_INFO*>(buffer.data());;
```
Comment (tigercosmos) [inline Pcap++/src/PcapLiveDevice.cpp:995]: nit: `;;`

**Comment by seladb** [comment]
@Dimi1010 did you have a chance to test it on Windows? We don't have tests for it unfortunately because it's not easy to run it in CI...

**Comment by Dimi1010** [comment]
> @Dimi1010 did you have a chance to test it on Windows? We don't have tests for it unfortunately because it's not easy to run it in CI...

I see. Haven't tested it yet, thought it was part of the CI. Will test it after I fix the merge conflicts.

**Comment by tigercosmos** [comment]
> @Dimi1010 did you have a chance to test it on Windows? We don't have tests for it unfortunately because it's not easy to run it in CI...

I am curious about why it is impossible to run it on CI?

**Comment by Dimi1010** [comment]
@seladb Tested it on Win10, using local adapter. All tests passed, with `TestRawSockets` receiving a non-critical failure. 
image

File: Pcap++/src/PcapLiveDevice.cpp (hunk 1 above)
Comment (Dimi1010) [reply Pcap++/src/PcapLiveDevice.cpp]: Fixed 8c2ec22

**Comment by seladb** [comment]
@Dimi1010 @tigercosmos I apologize, we do have tests for it which we run in CI as well: https://github.com/seladb/PcapPlusPlus/blob/6d17ca81fbc80dfd4fa3529b18ed81a753f192f3/Tests/Pcap%2B%2BTest/Tests/LiveDeviceTests.cpp#L164

I'm sorry for the confusion

**Review by seladb — APPROVED** [review]

**Pull request merged**